libnetwork/network.go

```diff
@@ -1223,7 +1220,7 @@ func (n *network) createEndpoint(name string, options ...EndpointOption) (*Endpo
 			ep.releaseAddress()
 		}
 	}()
-
+	n.ctrlr.networkLocker.Unlock(n.id)
 	if err = n.addEndpoint(ep); err != nil {
```

Comment:
Could you please add a code comment to document why it's sound to release the network lock before calling `n.addEndpoint`?

Reply:
> if n.addEndpoint doesn't hold the lock n.ctrlr.networkLocker.Lock(n.id), will create container quickly

It doesn't matter how much faster containers can be created if the change is buggy.

> I delete this lock commit https://github.com/moby/libnetwork/pull/1808/files , testcase TestServiceVIPReuse is still passed

That is not enough to prove that the lock is not needed. Removing the lock might introduce a race condition which the test cases do not catch.

There must be a reason why `networkLocker` is being held for the duration of `func (*network) createEndpoint`. What are those reasons? Which race conditions are being guarded against? How are you sure that those race conditions cannot happen after the changes in this PR? Please walk us through your reasoning.

Reply:
I'm not sure where are race conditions ,but I find  https://github.com/moby/moby/blame/master/libnetwork/endpoint.go#L836-L863  doesn't hold networkLocker
when I use docker rm a container.  I don't know why createEndpoint should hold the networkLocker